src/Gaffer/MetadataAlgo.cpp

```diff
@@ -218,6 +220,51 @@ void bookmarks( const Node *node, std::vector<NodePtr> &bookmarks )
 	}
 }
 
+void setViewerHotkey( Node *node, int hotkey )
+{
+	const ScriptNode *scriptNode = node->ancestor<ScriptNode>();
+	if( !scriptNode )
+	{
+		return;
+	}
+
+	for( RecursiveNodeIterator it( scriptNode ); !it.done(); ++it )
+	{
+		if( getViewerHotkey( it->get() ) == hotkey )
+		{
+			Metadata::deregisterValue( it->get(), g_viewerHotkeyName );
+			// \todo: in theory we could break here...
```

Comment:
Hm, I'm confused. You've probably thought this through already - I'll just talk my way through it here and maybe you can comment on what makes sense and what doesn't.

If we aren't allowed to change contents of references, it either means that 

1. We provide access to the bookmarks, but we block the numeric bookmark slots already taken up by nodes in referenced boxes OR
2. We'll have to ignore all numeric bookmarks within references in the Metadata API.

Is that correct? None of this is ideal, is it? Downsides of these are:

1. What if we we reference the same Box twice? That confuses the system.
2. Now people can't use the new mechanism for anything in References?

Is it necessary to keep References unmodified? Yes, we'd lose the metadata between saving and reopening a script, but maybe that's ok and we should be more concerned with providing a nice experience to users during a Gaffer session? Setting the same numeric bookmark again after reopening a script might be ok?

Reply:
You raise some interesting questions. I hadn't really thought things through, but I think I was assuming we'd ignore references when looking up numeric bookmarks, or ignore them unless some scope argument was passed (so if you were inside a Reference, the GraphEditor could jump to a bookmark already in there). So basically option 2 above. I presume people don't need to jump around in references much because they should be treated as black boxes other than for debugging. I also think it's very likely that folks publishing references will leave their hotkeys in there, and I don't think those are going to be very intuitive to the consumer of the reference. Maybe worth a quick chat with Murray et al to canvas opinions?

Reply:
I talked to Murray and we both agreed that doing (2) seems reasonable.